xbmc/games/agents/input/AgentInput.cpp

```diff
@@ -365,32 +451,42 @@ void CGameAgentManager::ProcessAgents(const PERIPHERALS::PeripheralVector& joyst
         if (!inputHandlingLock)
           inputHandlingLock = m_peripheralManager.RegisterEventLock();
 
-        // Reinitialize agent
-        agent.Deinitialize();
-        agent.Initialize();
+        // Reinitialize agent's controller
+        agentController.Deinitialize();
+        agentController.Initialize();
 
         SetChanged(true);
       }
     }
   }
 
-  // Remove expired agents
+  // Remove expired joysticks
   std::vector<std::string> expiredJoysticks;
-  for (const auto& agent : m_agents)
+  for (const auto& agentController : m_controllers)
   {
-    auto it = std::find_if(joysticks.begin(), joysticks.end(),
-                           [&agent](const PERIPHERALS::PeripheralPtr& joystick)
-                           { return agent->GetPeripheralLocation() == joystick->Location(); });
+    if (agentController->GetPeripheral()->Type() != PERIPHERALS::PERIPHERAL_JOYSTICK)
+      continue;
 
-    if (it == joysticks.end())
-      expiredJoysticks.emplace_back(agent->GetPeripheralLocation());
+    if (std::any_of(peripherals.begin(), peripherals.end(),
+                    [](const PERIPHERALS::PeripheralPtr& peripheral)
+                    { return peripheral->Type() == PERIPHERALS::PERIPHERAL_JOYSTICK; }))
+    {
+      auto it =
+          std::find_if(peripherals.begin(), peripherals.end(),
+                       [&agentController](const PERIPHERALS::PeripheralPtr& peripheral) {
+                         return agentController->GetPeripheralLocation() == peripheral->Location();
+                       });
```

Comment:
I don't think so, because the first if is `!= peripherals.end()` and the second if is `== peripherals.end()`. However, the first find can be factored further outside of the for loop. I pushed an update.